### PR TITLE
fixes to mailroom2 and added lesson 4 - Trigram Kata

### PR DESCRIPTION
**Pull request by mkdir-washington-edu**
Trigram_Kata.py

### CODE REVIEW

File: students/mkdir01/lesson04/Trigram_Kata.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python3
```
Comment (natasha-aleksandrova) [inline students/mkdir01/lesson04/Trigram_Kata.py]: note module names should also be snake_case -> `trigram_kata`

File: students/mkdir01/lesson04/Trigram_Kata.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            main = 0  # exit loop
+
+        counter += 1
+
+    write_file(file_name, new_story)
+
+
+def main():
+    read_file(input("input file name: "))
+    create_text(input("output file name: "), input("Minimum character length: "))
```
Comment (natasha-aleksandrova) [inline students/mkdir01/lesson04/Trigram_Kata.py]: in mailroom we dealt with a global data structure (something that is expected); however normally you wouldn't be dealing with global data structures.

here would be a good case for functions that return a value and pass into next function, like:
``` 
words = read_file_into_words(file_name)
create_new_text(words)
```

File: students/mkdir01/lesson04/Trigram_Kata.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            if tup in storage and elem not in storage[tup]:
+                storage[tup].append(elem)
+            else:
+                storage[tup] = [elem]
+
+
+def read_file(file_name):
+    f = open(file_name)
+    temp_file = f.read()
+    f.close()
```
Comment (natasha-aleksandrova) [inline students/mkdir01/lesson04/Trigram_Kata.py]: it is preferred to use `with` statement when working with files.

File: students/mkdir01/lesson04/Trigram_Kata.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    # take words and make into a list
+    add_to_storage(seperate_words(temp_file))
+
+
+def write_file(file_name, new_story):
+    f = open(file_name, 'w')
+    f.write(new_story)
+    f.close()
+    return
```
Comment (natasha-aleksandrova) [inline students/mkdir01/lesson04/Trigram_Kata.py]: `with` statement for working with files & no need for `return` statement.

File: students/mkdir01/lesson04/Trigram_Kata.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            return 0  # if the current tuple matches the initial tuple, main = 0
+        elif (tup[1], storage[tup][i]) in storage:
+            return detect_infinite_loop(tup[1], storage[tup][i])  # else keep searching the dict
+        else:
+            return 1  # we come to a point where we learn the loop is not infinite
+
+
+def create_key():
+    # choose a random key where the first word doesn't end in punctuation
+    tup = random.choice(list(storage.keys()))
```
Comment (natasha-aleksandrova) [inline students/mkdir01/lesson04/Trigram_Kata.py]: no need to call `.keys`

File: students/mkdir01/lesson04/Trigram_Kata.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    while tup[0] == '' or tup[0][-1] == ('.' or ',' or '"' or '!' or '?' or ';' or ':' or "'" or ')'):
```
Comment (natasha-aleksandrova) [inline students/mkdir01/lesson04/Trigram_Kata.py]: better -> `tup[0][-1] in ('.', ',', '!'):`

File: students/mkdir01/lesson04/Trigram_Kata.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+
+def look_for_pattern(temp_list):
+    for i in range(1, int(len(temp_list) / 2)):
+        # compares -1 to -2, then -1:-2 to -3:-4, etc...
+        if temp_list[-1:((-1 * i) - 1): -1] == temp_list[((-1 * i) - 1)::-1]:
+            global_loop_key0 = temp_list[(-1 * i)]  # first elem of looping list
+            global_loop_key1 = temp_list[(-1 * i) + 1]  # second elem of looping list
+            # copy the loop, from start to end (we've been traversing end to start)
+            return detect_infinite_loop((global_loop_key0, global_loop_key1))  # returns 0 if loop is found
```
Comment (natasha-aleksandrova) [inline students/mkdir01/lesson04/Trigram_Kata.py]: I'm not I sure I understand the infinite loop

**Comment by mkdir-washington-edu** [comment]
Thanks!

In a reductive example, the infinite loop happens if your dict has:
('Where', 'was') => 'I'
('I', 'was') => 'when'
('was', 'when') => 'I'
('when', 'I') => 'was'

Possible with shorter texts, so I wanted to build something that would
account for that.

**Pull request merged**